### PR TITLE
fix(install): emit "apm compile" hint for project-scope Gemini/agents/claude installs (#2057)

### LINKED ISSUE #2057: [BUG] Gemini instructions are not compiled into GEMINI.md when installing a local package
**Describe the bug**

When installing a local package with `apm install`, Gemini-specific instructions are not properly deployed, even when the package explicitly targets Gemini in `apm.yml`.

The installation generates or updates `GEMINI.md` with only a generation comment, but the actual instructions from the package are not included.

Also, `AGENTS.md` is not modified, which is expected in this case because the package target is `gemini`.

As a result, the package appears to be installed successfully, but Gemini does not receive the expected instructions.

**To Reproduce**

1. Create or use a local APM package with the following `apm.yml`:

```yaml
   name: gemini-ai-assets
   version: 1.0.0
   description: APM project for gemini-cli-config
   author: xxx

   # Which agent platforms to deploy to.
   # Resolution order: --target flag > this field > auto-detect from filesystem.
   # Accepted values: vscode, agents, copilot, claude, cursor, opencode, codex,
   # gemini, antigravity, windsurf, kiro, agent-skills, all
   targets:
     - gemini

   dependencies:
     apm: []
     mcp: []

   includes: auto

   scripts: {}

   compilation:
     agents_md:
       mode: managed_section
       start_marker: "<!-- apm:start -->"
       end_marker: "<!-- apm:end -->"
````

2. Run the installation command with the local package:

   ```bash
   apm install <local-package-path>
   ```

3. Check the generated files after installation.

4. Observe that:

   * `AGENTS.md` is not modified, which is expected because the target is `gemini`.
   * `GEMINI.md` is created or updated.
   * `GEMINI.md` only contains a generation comment.
   * The expected Gemini instructions from the package are missing.

**Expected behavior**

When installing a local package that explicitly targets Gemini using:

```yaml
targets:
  - gemini
```

`apm install` should include the package instructions in the generated `GEMINI.md`.

The generated `GEMINI.md` should contain the actual Gemini-specific instructions provided by the package, not only a generation comment.

Since the package targets Gemini, I do not expect `AGENTS.md` to be modified. However, I do expect `GEMINI.md` to contain the compiled instructions.

**Environment (please complete the following information):**

* OS: `Windows`
* APM Version: `0.24.0`

**Logs**

No explicit error is displayed during installation.

Observed generated `GEMINI.md` content:

```markdown
# GEMINI.md
<!-- Generated by APM CLI -->
<!-- Build ID: c1ee548a736b -->
<!-- APM Version: 0.23.1 -->

@./AGENTS.md
```

Install & compilation logs : 

```
front [ dev][⏱ 8s]
➜ apm install
[>] Installing dependencies from apm.yml...
[!] Could not fetch org policy from xxx (HTTP 410 fetching policy from xxx/xxx/.github); proceeding without policy enforcement. Retry, check 
connectivity, or use --no-policy to bypass.
[i] Targets: gemini  (source: apm.yml)
  [+] ./../gemini-cli-config/build/gemini-ai-assets-1.0.0 (local)
  |-- Skill integrated -> .agents/skills/
  |-- 2 skill(s) integrated -> .agents/skills/

[*] Installed 1 APM dependency in 1.6s.


front [ dev][⏱ 14s]
➜ apm compile
[*] Starting context compilation...
[i] Targets: gemini  (source: apm.yml)
[i] Compiling for GEMINI.md + .gemini/commands/ + .gemini/skills/ + .gemini/settings.json (MCP/hooks) - apm.yml target
[i] Generated GEMINI.md (imports AGENTS.md)
[+] Compilation completed successfully!
```

The expected Gemini instructions are not included.

**Additional context**

This issue occurs when deploying/installing a local package using `apm install`.

The package configuration explicitly targets Gemini through `apm.yml`, and `includes` is set to `auto`.

The installation appears to complete successfully, but the Gemini instructions are silently omitted from the generated `GEMINI.md`. This makes the issue difficult to detect unless the generated file is manually inspected.

The expected behavior would be for local package instructions targeting Gemini to be compiled into `GEMINI.md`.

```

### PR DESCRIPTION
**Pull request by sergio-sisternes-epam**
# fix(install): emit "apm compile" hint for project-scope Gemini/agents/claude installs (#2057)

## TL;DR

After `apm install <local-gemini-package>` on project scope, nothing told the user to run
`apm compile` — so GEMINI.md never received the installed instructions. This PR adds a
post-install hint for project-scope compile-only targets (gemini, agents, claude), mirroring
the user-scope hint introduced in #1632. No behavior change for existing flows.

> [!NOTE]
> Closes #2057. The compile path itself is not broken — `apm compile` always worked correctly;
> the gap was purely a missing UX signal after `apm install`.

## Problem (WHY)

- [x] `apm install <gemini-pkg>` on project scope exits 0 with no guidance. GEMINI.md remains
  a thin `@./AGENTS.md` stub; AGENTS.md is never updated. The user's dep instructions sit
  silently in `apm_modules/_local/<name>/.apm/instructions/` until they discover `apm compile`
  independently. Reported by Windows user on APM 0.24.0 in issue #2057.
- [x] `finalize.run()` already called `_hint_global_root_context()` for `InstallScope.USER`
  (added in #1632, [c4aa1c58](https://github.com/microsoft/apm/commit/c4aa1c58)), but the
  `InstallScope.PROJECT` branch had no equivalent. The routing was:
  `if USER: hint; # else: nothing`.
- [!] Copilot/cursor/kiro deploy instructions to per-file rules directories during `apm install`
  (no compile step needed). Gemini/agents/claude are compile-only by design: the `targets.py`
  TargetProfile for gemini has no `"instructions"` key. This difference is not surfaced to the user.

Why these matter: the DevX promise is that APM is ["pragmatic as npm"](https://github.com/microsoft/apm/blob/main/MANIFESTO.md) — first-run flows must not silently succeed while leaving configuration incomplete.

## Approach (WHAT)

| # | Fix |
|---|-----|
| 1 | Add `_has_dep_instruction_files(ctx)` — lightweight `rglob` scan of `apm_modules/` for `*.instructions.md` under `.apm/instructions/` subtrees. |
| 2 | Add `_hint_project_compile_needed(ctx)` — fires when a compile-only family target (gemini, agents, claude) is active AND dep instruction files exist in `apm_modules/`; emits one info line pointing at `apm compile`. |
| 3 | Update `finalize.run()` routing: `if USER → _hint_global_root_context; elif PROJECT → _hint_project_compile_needed`. `None`-scope installs remain silent (intentional — see Trade-offs). |

## Implementation (HOW)

- **`src/apm_cli/install/phases/finalize.py`** — Two new functions added before `_hint_global_root_context`. `_has_dep_instruction_files` scans `ctx.apm_modules_dir` (or `project_root/apm_modules`) via `rglob` with a path-parts guard. `_hint_project_compile_needed` reuses the existing `_ROOT_CONTEXT_ONLY_FAMILIES` and `_ROOT_CONTEXT_HINT_EXCLUDED_TARGETS` constants; guards on `dry_run`, `installed_count`, and `for_scope(user_scope=False)`. `run()` routing updated from a bare `if USER` to `if USER / elif PROJECT`.

- **`tests/unit/install/phases/test_finalize_user_compile.py`** — Module docstring and `test_project_scope_no_hint` updated to reflect the new routing. 11 new tests added in `TestHintProjectCompileNeeded`: fire condition (gemini + instructions), six no-fire conditions (copilot, no files, dry-run, zero-count, excluded target name), and a `run()`-level routing assertion.

- **`tests/unit/install/test_finalize_phase.py`** — `_FakeCtx` dataclass gained `dry_run: bool = False`, `targets: list = []`, and `apm_modules_dir: Any = None` so existing `run()` tests continue to pass when `_hint_project_compile_needed` is now reached for PROJECT-scoped contexts.

## Diagrams

Legend: `finalize.run()` post-install hint routing — dashed borders mark the two new nodes added by this PR.

```mermaid
flowchart LR
    run["finalize.run()"] --> scope{"ctx.scope?"}
    scope -->|"USER"| global["_hint_global_root_context()"]
    scope -->|"PROJECT"| project["_hint_project_compile_needed()"]
    scope -->|"None / other"| noop["(no hint)"]
    project --> check1{"compile-only target active?"}
    check1 -->|"No"| noop
    check1 -->|"Yes"| check2{"dep .instructions.md in apm_modules/?"}
    check2 -->|"No"| noop
    check2 -->|"Yes"| emit["emit: Run 'apm compile' hint on logger"]

    classDef new stroke-dasharray: 5 5;
    class project,check1,check2,emit new;
```

## Trade-offs

- **Hint-only, no auto-compile.** Chose hint; rejected auto-triggering `apm compile` because
  compilation is a separate, explicit user action — consistent with the user-scope model
  from #1632 and the PROSE principle
  ["favor small, chainable primitives."](https://danielmeppiel.github.io/awesome-ai-native/docs/prose/)
- **`elif PROJECT` not bare `else`.** `None`-scope contexts are transitional / test-only; silently
  hinting there would produce noise in edge cases without confirmed user benefit.
- **Filesystem scan at end of install.** The `rglob` is bounded to `apm_modules/` and returns on the
  first match, keeping it O(1) for the common case. No discovery package is imported; imports stay lazy.

## Benefits

1. Users who install a local Gemini/agents/claude package immediately see
   `[i] Instructions installed for Gemini CLI. Run 'apm compile' to update AGENTS.md / GEMINI.md.`
   — no silent success, no manual discovery of `apm compile` required.
2. Zero behavior change for existing user-scope installs, copilot/cursor/kiro
   project installs, and dry-run installs.
3. 1,777 unit tests pass; full lint mirror (ruff + pylint R0801 + auth-signal check) is green.

## Validation

<details><summary>Lint mirror (all green)</summary>

```
$ uv run --extra dev ruff check src/ tests/ && uv run --extra dev ruff format --check src/ tests/ \
  && uv run --extra dev python -m pylint --disable=all --enable=R0801 \
     --min-similarity-lines=10 --fail-on=R0801 src/apm_cli/ \
  && bash scripts/lint-auth-signals.sh

All checks passed!
1417 files already formatted

Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)

[+] auth-signal lint clean
```

</details>

<details><summary>Unit tests — 1,777 passed</summary>

```
$ uv run --extra dev pytest tests/unit/install/ -q
1777 passed in 18.59s
```

New hint tests specifically:

```
$ uv run --extra dev pytest tests/unit/install/phases/test_finalize_user_compile.py \
    tests/unit/install/test_finalize_phase.py -q
41 passed in 0.30s
```

</details>

<details><summary>Smoke test — hint fires on install</summary>

```
[i] Instructions installed for gemini. Run 'apm compile' to update AGENTS.md / GEMINI.md.
[*] Installed 1 APM dependency in 0.1s.

compile hint shown: True
```

</details>

### Scenario Evidence

| # | Scenario (user promise) | Principle(s) | Test(s) proving it | Type |
|---|-------------------------|--------------|--------------------|------|
| 1 | After `apm install <gemini-pkg>`, user immediately sees a "Run 'apm compile'" hint — no silent success | DevX | `tests/unit/install/phases/test_finalize_user_compile.py::TestHintProjectCompileNeeded::test_hint_fires_for_compile_only_target_with_instruction_files`<br/>`test_run_calls_project_hint_for_project_scope` (regression-trap for #2057) | unit |
| 2 | Copilot/cursor/kiro project installs do NOT show the hint (no spurious guidance for targets that deploy instructions natively) | DevX, Multi-harness support | `test_hint_not_fired_for_copilot_target`<br/>`test_hint_not_fired_for_excluded_target_name` | unit |
| 3 | Dry-run and no-op (`installed_count == 0`) installs remain silent | DevX, Governed by policy | `test_hint_not_fired_on_dry_run`<br/>`test_hint_not_fired_when_nothing_installed` | unit |
| 4 | User-scope installs still see their existing `apm compile -g` hint (no regression) | DevX | `tests/unit/install/phases/test_finalize_user_compile.py::TestFinalizeRunIntegration::test_user_scope_hint_called` | unit |

## How to test

- [ ] In a temp dir, create a Gemini-target local package with a `.apm/instructions/rules.instructions.md`. Run `apm install ./package` and confirm the output contains `Run 'apm compile'`.
- [ ] Run `apm compile` in the same dir and confirm AGENTS.md contains the instruction content. GEMINI.md should contain `@./AGENTS.md`.
- [ ] Repeat with a Copilot-target package; confirm no compile hint appears in the output.
- [ ] Run `apm install --dry-run ./package` with the Gemini package; confirm no hint.
- [ ] Run `uv run --extra dev pytest tests/unit/install/phases/test_finalize_user_compile.py -q` — expect 20 passed.

Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>

### CODE REVIEW

File: src/apm_cli/install/phases/finalize.py [hunk 1]
```diff
@@ -28,6 +28,82 @@
...
+
+    if not _has_dep_instruction_files(ctx):
+        return
+
+    if ctx.logger:
+        targets = ", ".join(target_names)
+        message = (
+            f"Instructions installed for {targets}. "
+            "Run 'apm compile' to update AGENTS.md / GEMINI.md."
+        )
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply src/apm_cli/install/phases/finalize.py:103]: Fixed in c9ce69cf. The message now reads "Run 'apm compile' to update root context files." — which covers all three compile families (gemini, agents, claude) and matches the docstring language.


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

**Comment by danielmeppiel** [comment]
## APM Review Panel: `needs_rework`

> Closes the silent first-run gap for project-scope Gemini/agents/claude installs by emitting an `apm compile` hint when compile-only targets carry dependency instructions.

cc @sergio-sisternes-epam -- a fresh advisory pass is ready for your review.

Four independent panelists converged on one bounded message fix: the emitted hint says "root context files" where concrete artifact names would make the next step immediately verifiable. The message and its regression assertion should agree.

The supply-chain review also found that `Path.rglob` follows symlinked directories on supported Python versions. The scan is read-only and the current impact is a false-positive hint, but the new path should still honor APM's containment invariant. Add a containment guard and a symlink regression test.

The remaining signals are documentation and scan-semantics follow-ups. Existing focused and integration tests pass; no reviewer found a happy-path behavior regression.

**Dissent.** The performance reviewer proposed replacing the scan with staged state. The measured risk is currently noise, so that larger refactor is not warranted here.

**Aligned with:** pragmatic as npm through an explicit next step; multi-harness support through target-aware guidance; secure by default through path containment.

**Growth signal.** This removes a silent-success first-run failure. Naming the output artifacts makes the terminal hint easier to understand and share.

### Panel summary

| Persona | B | R | N | Takeaway |
|---|---|---|---|---|
| Python Architect | 0 | 2 | 2 | Clean scope dispatch; align message evidence and document conservative scan semantics. |
| CLI Logging Expert | 0 | 1 | 1 | CommandLogger use is correct; name the concrete output artifacts. |
| DevX UX Expert | 0 | 1 | 2 | The hint fires correctly; make its wording actionable and guarded. |
| Supply Chain Security Expert | 0 | 1 | 0 | Add containment around the read-only recursive scan. |
| OSS Growth Hacker | 0 | 1 | 1 | Close the first-run gap with concrete artifact wording. |
| Doc Writer | 0 | 3 | 1 | Document the new project-scope hint in the user corpus. |
| Test Coverage Expert | 0 | 0 | 1 | Coverage is stronger than the PR body reports. |
| Performance Expert | 0 | 1 | 2 | Current scan cost is small; document its worst-case behavior. |

> B = blocking-severity findings, R = recommended, N = nits.
> Counts are signal strength, not gates. The maintainer ships.

### Top 4 follow-ups

1. **[Python Architect / CLI Logging Expert / DevX UX Expert / OSS Growth Hacker]** Replace `root context files` with concrete artifact names and tighten the assertion -- four reviewers independently identified this bounded first-run UX improvement.
2. **[Supply Chain Security Expert]** Add a containment guard around the recursive scan and a symlink regression test -- supported Python versions can follow symlinked directories.
3. **[Doc Writer]** Update the install reference, consumer install guide, and first-package guide -- the new terminal behavior currently has no documentation home.
4. **[Python Architect]** Document the intentional all-`apm_modules` scan -- historical dependency instructions can conservatively trigger the hint.

### Architecture

```mermaid
flowchart TD
    A["finalize.run"] --> B{"Install scope"}
    B -->|USER| C["Existing global compile hint"]
    B -->|PROJECT| D["Project compile hint"]
    D --> E{"Compile-only target active"}
    E -->|No| F["No hint"]
    E -->|Yes| G["Scan dependency instructions"]
    G -->|Found| H["CommandLogger info hint"]
    G -->|Missing| F
```

### Recommendation

Fold the message/assertion update and path-containment regression test before shipping. Then fold the directly affected documentation and clarify the conservative scan semantics.

---

<details>
<summary>Full per-persona findings</summary>

#### Python Architect

- **[recommended]** PR body evidence and emitted message disagree. Tighten the assertion after choosing concrete wording.
- **[recommended]** The scan includes historical dependency instructions. Document that conservative intent.
- **[nit]** Align project- and user-scope hint sentence structure.
- **[nit]** Explain the valid `for_scope()` `None` case.

#### CLI Logging Expert

- **[recommended]** `root context files` is abstract at the point where users need a concrete next step.
- **[nit]** PR smoke output does not match the final message.

#### DevX UX Expert

- **[recommended]** Name the artifacts users can inspect after compilation and guard that wording in the test.
- **[nit]** PR smoke evidence drifted from the code.
- **[nit]** The user- and project-scope messages use different sentence structures.

#### Supply Chain Security Expert

- **[recommended]** `rglob` can follow symlinked directories on supported Python versions. Resolve each candidate through the repository's path-containment helper before accepting it.

#### OSS Growth Hacker

- **[recommended]** Concrete artifact names make the first-run next step easier to act on.
- **[nit]** Add a concise changelog entry for the community-reported fix.

#### Auth Expert -- inactive

The touched files do not change tokens, credentials, hosts, or authentication.

#### Doc Writer

- **[recommended]** Extend `reference/cli/install.md` with the project-scope compile hint.
- **[recommended]** Extend `consumer/install-packages.md` with the install-then-compile flow.
- **[recommended]** Correct `getting-started/first-package.md` for compile-only targets.
- **[nit]** Align PR-body terminology with the final message.

#### Test Coverage Expert

- **[nit]** The PR body omits the passing integration-with-fixtures regression trap for Gemini install output.

#### Performance Expert

- **[recommended]** Document that the scan is worst-case O(total files in `apm_modules`).
- **[nit]** A scaling floor test would guard future scan regressions.
- **[nit]** Avoid unconditional path resolution if containment handling makes it unnecessary.

</details>

<sub>This panel is advisory. It does not block merge. Re-apply the `panel-review` label after addressing feedback to re-run.</sub>

**Comment by danielmeppiel** [comment]
Thank you for the original work on this fix. To land it promptly we have opened a superseding PR (#2293) under microsoft/apm that preserves your authorship via commit trailers and resolves the follow-ups surfaced by the apm-review-panel pass.

Closing this PR in favor of #2293. Your contribution is credited on every cherry-picked commit; the superseding PR's body links back here. Please do raise concerns on the superseding PR if the changes diverge from your intent -- we want your sign-off too.
